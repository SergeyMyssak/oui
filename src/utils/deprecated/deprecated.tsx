/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect } from 'react';

export const getDeprecatedMessage = (message: string): string =>
  `[DEPRECATED] ${message}`;

export const deprecated = (message: string) => {
  return <T extends React.ComponentType<any>>(Component: T): T => {
    const DeprecatedWrapper = (props: React.ComponentProps<T>) => {
      useEffect(() => {
        console.warn(getDeprecatedMessage(message));
      }, []);

      return <Component {...props} />;
    };

    Object.defineProperty(DeprecatedWrapper, 'name', {
      value: Component.displayName || Component.name,
    });

    return DeprecatedWrapper as T;
  };
};
