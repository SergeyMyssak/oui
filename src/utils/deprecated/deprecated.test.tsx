/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { mount, render } from 'enzyme';
import { deprecated, getDeprecatedMessage } from './deprecated';

describe('deprecated', () => {
  const warning = 'This component is deprecated in favor of another.';

  it('should console warning', () => {
    console.warn = jest.fn();

    const Component = () => <div />;
    const DeprecatedComponent = deprecated(warning)(Component);
    mount(<DeprecatedComponent />);

    expect(console.warn).toHaveBeenCalledWith(getDeprecatedMessage(warning));
  });

  it('should render component', () => {
    console.warn = jest.fn();

    const Component = () => <div />;
    const DeprecatedComponent = deprecated(warning)(Component);

    const component = render(<DeprecatedComponent />);
    expect(component).toMatchSnapshot();
  });

  it('should properly name DeprecatedWrapper function', () => {
    const Component = () => <div />;
    const DeprecatedComponent = deprecated(warning)(Component);

    expect(DeprecatedComponent.name).toEqual('Component');
  });
});
