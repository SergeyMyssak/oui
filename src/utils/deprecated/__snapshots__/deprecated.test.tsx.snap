// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deprecated should render component 1`] = `<div />`;
